### PR TITLE
Dashboard As Code for Reconcile

### LINKED ISSUE #708: [TECH DEBT]: Use LSQL for Dashboards-as-code instead of unreadable JSON files
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Category of feature request

Other

### Problem statement

Use LSQL for dashboards-as-code

### Proposed Solution

See docs at https://github.com/databrickslabs/lsql/blob/main/docs/dashboards.md
See UCX codebase for example usage

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by sundarshankar89**
closes #708 

The existing dashboard has been broken into 2.
For features unsupported by LSQL, we are using overrides to implement the required widgets.

Conditional formatting depends on https://github.com/databrickslabs/lsql/pull/299

Screenshots:

REMORPH  Reconciliation Metrics
REMORPH  Aggregate Reconciliation Metrics

-- co-authored by @bishwajit-db and @sundarshankar89

### CODE REVIEW

File: src/databricks/labs/remorph/resources/reconcile/dashboards/failed_table_list.sql [hunk 1]
```diff
@@ -0,0 +1,26 @@
+select
+main.recon_id,
+main.source_type,
+main.report_type,
+main.source_table.`catalog` as source_catalog,
+main.source_table.`schema` as source_schema,
+main.source_table.table_name as source_table_name,
+CASE
```
Comment (vijaypavann-db) [inline src/databricks/labs/remorph/resources/reconcile/dashboards/failed_table_list.sql]: Please update the `CASE..WHEN..END` statement to `if(cond, exp1, exp2)`... to simplify the statement and improve readability.

```suggestion
IF(ISNULL(MAIN.SOURCE_TABLE.CATALOG)
    , CONCAT_WS('.', MAIN.SOURCE_TABLE.SCHEMA, MAIN.SOURCE_TABLE.TABLE_NAME)
    , CONCAT_WS('.', MAIN.SOURCE_TABLE.CATALOG, MAIN.SOURCE_TABLE.SCHEMA, MAIN.SOURCE_TABLE.TABLE_NAME)) AS source_table,
```

File: src/databricks/labs/remorph/resources/reconcile/dashboards/recon_details_pivot.sql [hunk 2]
```diff
@@ -0,0 +1,22 @@
+with tmp as (select recon_table_id, inserted_ts ,recon_type, explode(data) as data,
+row_number() over(partition by recon_table_id,recon_type order by recon_table_id) as rn
+from IDENTIFIER(:catalog || '.' || :schema || '.details' )
+where recon_type != 'schema')
+select main.recon_id,
+main.source_table.`catalog` as source_catalog,
+main.source_table.`schema` as source_schema,
+main.source_table.table_name as source_table_name,
+CASE
```
Comment (vijaypavann-db) [inline src/databricks/labs/remorph/resources/reconcile/dashboards/recon_details_pivot.sql]: Please update the `CASE..WHEN..END` statement to `if(cond, exp1, exp2)`... to simplify the statement and improve readability.

```suggestion
IF(ISNULL(MAIN.SOURCE_TABLE.CATALOG)
    , CONCAT_WS('.', MAIN.SOURCE_TABLE.SCHEMA, MAIN.SOURCE_TABLE.TABLE_NAME)
    , CONCAT_WS('.', MAIN.SOURCE_TABLE.CATALOG, MAIN.SOURCE_TABLE.SCHEMA, MAIN.SOURCE_TABLE.TABLE_NAME)) AS source_table,
```

File: src/databricks/labs/remorph/resources/reconcile/dashboards/success_table_list.sql [hunk 3]
```diff
@@ -0,0 +1,20 @@
+select
+main.recon_id,
+main.source_type,
+main.report_type,
+CASE
```
Comment (vijaypavann-db) [inline src/databricks/labs/remorph/resources/reconcile/dashboards/success_table_list.sql]: Please update the `CASE..WHEN..END` statement to `if(cond, exp1, exp2)`... to simplify the statement and improve readability.

```suggestion
IF(ISNULL(MAIN.SOURCE_TABLE.CATALOG)
    , CONCAT_WS('.', MAIN.SOURCE_TABLE.SCHEMA, MAIN.SOURCE_TABLE.TABLE_NAME)
    , CONCAT_WS('.', MAIN.SOURCE_TABLE.CATALOG, MAIN.SOURCE_TABLE.SCHEMA, MAIN.SOURCE_TABLE.TABLE_NAME)) AS source_table,
```

File: src/databricks/labs/remorph/resources/reconcile/dashboards/recon_main.sql [hunk 4]
```diff
@@ -0,0 +1,31 @@
+select
+main.recon_id,
+main.source_type,
+main.report_type,
+main.source_table.`catalog` as source_catalog,
+main.source_table.`schema` as source_schema,
+main.source_table.table_name as source_table_name,
+CASE
```
Comment (vijaypavann-db) [inline src/databricks/labs/remorph/resources/reconcile/dashboards/recon_main.sql]: Please update the `CASE..WHEN..END` statement to `if(cond, exp1, exp2)`... to simplify the statement and improve readability.

```suggestion
IF(ISNULL(MAIN.SOURCE_TABLE.CATALOG)
    , CONCAT_WS('.', MAIN.SOURCE_TABLE.SCHEMA, MAIN.SOURCE_TABLE.TABLE_NAME)
    , CONCAT_WS('.', MAIN.SOURCE_TABLE.CATALOG, MAIN.SOURCE_TABLE.SCHEMA, MAIN.SOURCE_TABLE.TABLE_NAME)) AS source_table,
```

File: src/databricks/labs/remorph/resources/reconcile/dashboards/recon_details_schema.sql [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+main.source_table.`schema` as source_schema,
+main.source_table.table_name as source_table_name,
+CASE
+        WHEN COALESCE(MAIN.SOURCE_TABLE.CATALOG, '') <> '' THEN CONCAT(MAIN.SOURCE_TABLE.CATALOG, '.', MAIN.SOURCE_TABLE.SCHEMA, '.', MAIN.SOURCE_TABLE.TABLE_NAME)
+        ELSE CONCAT(MAIN.SOURCE_TABLE.SCHEMA, '.', MAIN.SOURCE_TABLE.TABLE_NAME)
+    END AS source_table,
+main.target_table.`catalog` as target_catalog,
+main.target_table.`schema` as target_schema,
+main.target_table.table_name as target_table_name,
+CONCAT(MAIN.TARGET_TABLE.CATALOG, '.', MAIN.TARGET_TABLE.SCHEMA, '.', MAIN.TARGET_TABLE.TABLE_NAME) AS target_table,
```
Comment (vijaypavann-db) [inline src/databricks/labs/remorph/resources/reconcile/dashboards/recon_details_schema.sql]: Why are we repeating the same `Target Catalog, Schema, Table` Information again?

```suggestion
CONCAT_WS('.', MAIN.TARGET_TABLE.CATALOG, MAIN.TARGET_TABLE.SCHEMA, MAIN.TARGET_TABLE.TABLE_NAME) AS target_table,
```
Or

```suggestion
CONCAT_WS('.', TARGET_CATALOG, TARGET_SCHEMA, TARGET_TABLE_NAME) AS target_table,
```

**Review by vijaypavann-db — CHANGES_REQUESTED** [review]

File: src/databricks/labs/remorph/resources/reconcile/dashboards/dashboard.yml
[the remark this answers is not in the document]
Comment (sundarshankar89) [reply src/databricks/labs/remorph/resources/reconcile/dashboards/dashboard.yml:1]: Will remove this

File: src/databricks/labs/remorph/deployment/dashboard.py [hunk 6]
```diff
@@ -1,82 +1,105 @@
...
-        logger.info(f"Dashboard deployed with dashboard_id {dashboard.dashboard_id}")
-        logger.info(f"Dashboard URL: {self._ws.config.host}/sql/dashboardsv3/{dashboard.dashboard_id}")
+    def _handle_existing_dashboard(self, dashboard_id: str, display_name: str, parent_path: str) -> str | None:
+        try:
+            dashboard = self._ws.lakeview.get(dashboard_id)
+            if dashboard.lifecycle_state is None:
+                raise NotFound(f"Dashboard life cycle state: {display_name} ({dashboard_id})")
+            if dashboard.lifecycle_state == LifecycleState.TRASHED:
+                logger.info(f"Recreating trashed dashboard: {display_name} ({dashboard_id})")
+                return None  # Recreate the dashboard if it is trashed (manually)
```
[the remark this answers is not in the document]
Comment (sundarshankar89) [reply src/databricks/labs/remorph/deployment/dashboard.py:39]: will breakdown the functions

File: src/databricks/labs/remorph/resources/reconcile/dashboards/0_1_recon_main.sql [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+main.source_table.`schema` as source_schema,
+main.source_table.table_name as source_table_name,
+CASE
+        WHEN COALESCE(MAIN.SOURCE_TABLE.CATALOG, '') <> '' THEN CONCAT(MAIN.SOURCE_TABLE.CATALOG, '.', MAIN.SOURCE_TABLE.SCHEMA, '.', MAIN.SOURCE_TABLE.TABLE_NAME)
+        ELSE CONCAT(MAIN.SOURCE_TABLE.SCHEMA, '.', MAIN.SOURCE_TABLE.TABLE_NAME)
+    END AS source_table,
+main.target_table.`catalog` as target_catalog,
+main.target_table.`schema` as target_schema,
+main.target_table.table_name as target_table_name,
+CONCAT(MAIN.TARGET_TABLE.CATALOG, '.', MAIN.TARGET_TABLE.SCHEMA, '.', MAIN.TARGET_TABLE.TABLE_NAME) AS target_table,
```
[the remark this answers is not in the document]
Comment (sundarshankar89) [reply src/databricks/labs/remorph/resources/reconcile/dashboards/0_1_recon_main.sql:15]: Followed standard SQL formatting now expect for keywords rest all will be in smaller case.

**Comment by sundarshankar89** [comment]
Blocked by 

LSQL Features:

- https://github.com/databrickslabs/lsql/issues/173
- https://github.com/databrickslabs/lsql/issues/246
- https://github.com/databrickslabs/lsql/issues/124
- https://github.com/databrickslabs/lsql/issues/116
- https://github.com/databrickslabs/lsql/issues/115

File: src/databricks/labs/remorph/resources/reconcile/dashboards/recon_details_pivot.sql [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+        WHEN COALESCE(MAIN.SOURCE_TABLE.CATALOG, '') <> '' THEN CONCAT(MAIN.SOURCE_TABLE.CATALOG, '.', MAIN.SOURCE_TABLE.SCHEMA, '.', MAIN.SOURCE_TABLE.TABLE_NAME)
+        ELSE CONCAT(MAIN.SOURCE_TABLE.SCHEMA, '.', MAIN.SOURCE_TABLE.TABLE_NAME)
+    END AS source_table,
+main.target_table.`catalog` as target_catalog,
+main.target_table.`schema` as target_schema,
+main.target_table.table_name as target_table_name,
+CONCAT(MAIN.TARGET_TABLE.CATALOG, '.', MAIN.TARGET_TABLE.SCHEMA, '.', MAIN.TARGET_TABLE.TABLE_NAME) AS target_table,
```
[the remark this answers is not in the document]
Comment (sundarshankar89) [reply src/databricks/labs/remorph/resources/reconcile/dashboards/recon_details_pivot.sql]: `target_table` is repeated for display inside the visualisation

File: src/databricks/labs/remorph/deployment/dashboard.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (sundarshankar89) [reply src/databricks/labs/remorph/deployment/dashboard.py:40]: Can't break down the functions this will be the exception to the return statement rule.

**Comment by sundarshankar89** [comment]
@nfx  why is it moved to in progress? it is blocked by lsql missing features.

File: src/databricks/labs/remorph/deployment/dashboard.py [hunk 9]
```diff
@@ -1,82 +1,107 @@
...
+
+        except (NotFound, InvalidParameterValue):
+            logger.info(f"Recovering invalid dashboard: {display_name} ({dashboard_id})")
+            try:
+                dashboard_path = f"{parent_path}/{display_name}.lvdash.json"
+                self._ws.workspace.delete(dashboard_path)  # Cannot recreate dashboard if file still exists
+                logger.debug(f"Deleted dangling dashboard {display_name} ({dashboard_id}): {dashboard_path}")
+            except NotFound:
+                pass
+            return None  # Recreate the dashboard if it's reference is corrupted (manually)
```
Comment (nfx) [inline src/databricks/labs/remorph/deployment/dashboard.py:50]: don't return None, bubble up NotFound and handle it in the client code.

here we have some anti-pattern logic:
```
try: 
  if not ...: raise NotFound
  return None
except NotFound:
  try:
     if not ...: raise NotFound
  except NotFound:
     return None
```

File: src/databricks/labs/remorph/deployment/dashboard.py [hunk 10]
```diff
@@ -1,82 +1,107 @@
...
+        metadata.display_name = self._name_with_prefix(metadata.display_name)
+
+        dashboard_id = self._install_state.dashboards.get(name)
+        if dashboard_id is not None:
+            dashboard_id = self._handle_existing_dashboard(dashboard_id, metadata.display_name, parent_path)
+
+        # dashboard_data = self._substitute_params(dashboard_file, parameters or {})
+        dashboard_id = self._update_or_create_dashboard(name, dashboard_id, metadata, parent_path)
+        logger.info(f"Dashboard deployed with dashboard_id {dashboard_id}")
+        logger.info(f"Dashboard URL: {self._ws.config.host}/sql/dashboardsv3/{dashboard_id}")
```
Comment (nfx) [inline src/databricks/labs/remorph/deployment/dashboard.py:83]: compress two log lines to one

**Review by nfx — CHANGES_REQUESTED** [review]
minor changes

File: src/databricks/labs/remorph/deployment/dashboard.py [hunk 11]
```diff
@@ -1,82 +1,107 @@
...
+        folder: Path,
+        config: ReconcileMetadataConfig,
+    ):
+        """Create a dashboard from Queries inside folder"""
+        logger.info(f"Deploying dashboard {name} from {folder}")
+        parent_path = f"{self._installation.install_folder()}/dashboards"
+        try:
+            self._ws.workspace.mkdirs(parent_path)
+        except ResourceAlreadyExists:
+            pass
```
Comment (vijaypavann-db) [inline src/databricks/labs/remorph/deployment/dashboard.py:65]: Adding a logger message would be good here

File: src/databricks/labs/remorph/deployment/dashboard.py [hunk 12]
```diff
@@ -1,82 +1,107 @@
...
+            database_to_replace="remorph_schema",
+        )
+
+        metadata.display_name = self._name_with_prefix(metadata.display_name)
+
+        dashboard_id = self._install_state.dashboards.get(name)
+        if dashboard_id is not None:
+            dashboard_id = self._handle_existing_dashboard(dashboard_id, metadata.display_name, parent_path)
+
+        # dashboard_data = self._substitute_params(dashboard_file, parameters or {})
```
Comment (vijaypavann-db) [inline src/databricks/labs/remorph/deployment/dashboard.py:80]: Pls remove the commented out code

File: src/databricks/labs/remorph/resources/reconcile/dashboards/reconciliation_metrics/05_0_summary_table.sql [hunk 13]
```diff
@@ -0,0 +1,32 @@
+/* --title 'Summary Table' --width 6 --height 6 */
+SELECT main.recon_id,
+       main.source_type,
+       main.report_type,
+       main.source_table.`catalog` AS source_catalog,
+       main.source_table.`schema` AS source_schema,
+       main.source_table.table_name AS source_table_name,
+       CASE
+           WHEN COALESCE(main.source_table.catalog, '') <> '' THEN CONCAT(main.source_table.catalog, '.', main.source_table.schema, '.', main.source_table.table_name)
```
Comment (vijaypavann-db) [inline src/databricks/labs/remorph/resources/reconcile/dashboards/reconciliation_metrics/05_0_summary_table.sql:9]: This case statement can be converted as below, avoiding Unnecessary use of `Coalesece`
```
 IF(
                     ISNULL(source_catalog),
                     CONCAT_WS('.', source_schema, source_table_name),
                     CONCAT_WS(
                             '.',
                             source_catalog,
                             source_schema,
                             source_table_name
                     )
             ) AS source_table,
```

File: src/databricks/labs/remorph/resources/reconcile/dashboards/reconciliation_metrics/07_0_schema_details_table.sql [hunk 14]
```diff
@@ -0,0 +1,36 @@
...
+    WHERE
+        recon_type = 'schema'
+)
+SELECT
+    main.recon_id,
+    main.source_table.`catalog` AS source_catalog,
+    main.source_table.`schema` AS source_schema,
+    main.source_table.table_name AS source_table_name,
+    CASE
+        WHEN COALESCE(main.source_table.catalog, '') <> '' THEN CONCAT(main.source_table.catalog, '.', main.source_table.schema, '.', main.source_table.table_name)
```
Comment (vijaypavann-db) [inline src/databricks/labs/remorph/resources/reconcile/dashboards/reconciliation_metrics/07_0_schema_details_table.sql:18]: This case statement can be converted as below, avoiding the Unnecessary use of `Coalesece`
```
 IF(
                     ISNULL(source_catalog),
                     CONCAT_WS('.', source_schema, source_table_name),
                     CONCAT_WS(
                             '.',
                             source_catalog,
                             source_schema,
                             source_table_name
                     )
             ) AS source_table,
```

File: src/databricks/labs/remorph/resources/reconcile/dashboards/reconciliation_metrics/11_0_recon_details_pivot.sql [hunk 15]
```diff
@@ -0,0 +1,34 @@
...
+    WHERE
+        recon_type != 'schema'
+)
+SELECT
+    main.recon_id AS dd_recon_id,
+    main.source_table.`catalog` AS source_catalog,
+    main.source_table.`schema` AS source_schema,
+    main.source_table.table_name AS source_table_name,
+    CASE
+        WHEN COALESCE(main.source_table.catalog, '') <> '' THEN CONCAT(main.source_table.catalog, '.', main.source_table.schema, '.', main.source_table.table_name)
```
Comment (vijaypavann-db) [inline src/databricks/labs/remorph/resources/reconcile/dashboards/reconciliation_metrics/11_0_recon_details_pivot.sql:20]: This case statement can be converted as below, avoiding the Unnecessary use of `Coalesece`
```
 IF(
                     ISNULL(source_catalog),
                     CONCAT_WS('.', source_schema, source_table_name),
                     CONCAT_WS(
                             '.',
                             source_catalog,
                             source_schema,
                             source_table_name
                     )
             ) AS source_table,
```

File: src/databricks/labs/remorph/resources/reconcile/dashboards/reconciliation_metrics/15_1_failed_targets.sql [hunk 16]
```diff
@@ -0,0 +1,10 @@
+/* --title 'Unique target tables failed' --width 2 */
+SELECT
+    CONCAT(main.target_table.catalog, '.', main.target_table.schema, '.', main.target_table.table_name) AS t_table,
```
Comment (vijaypavann-db) [inline src/databricks/labs/remorph/resources/reconcile/dashboards/reconciliation_metrics/15_1_failed_targets.sql:3]: `CONCAT_WS` could be used here

File: src/databricks/labs/remorph/resources/reconcile/dashboards/reconciliation_metrics/15_2_successful_targets.sql [hunk 17]
```diff
@@ -0,0 +1,10 @@
+/* --title 'Unique target tables successful' --width 2 */
+SELECT
+    CONCAT(main.target_table.catalog, '.', main.target_table.schema, '.', main.target_table.table_name) AS t_table,
```
Comment (vijaypavann-db) [inline src/databricks/labs/remorph/resources/reconcile/dashboards/reconciliation_metrics/15_2_successful_targets.sql:3]: `CONCAT_WS` could be used here

File: src/databricks/labs/remorph/resources/reconcile/dashboards/reconciliation_metrics/17_0_mismatched_records.sql [hunk 18]
```diff
@@ -0,0 +1,14 @@
+/* --title 'Mismatched Records' --width 3 */
+SELECT
+    main.recon_id,
+    CONCAT(main.target_table.catalog, '.', main.target_table.schema, '.', main.target_table.table_name) AS target_table,
```
Comment (vijaypavann-db) [inline src/databricks/labs/remorph/resources/reconcile/dashboards/reconciliation_metrics/17_0_mismatched_records.sql:4]: `CONCAT_WS` could be used here

File: src/databricks/labs/remorph/resources/reconcile/dashboards/reconciliation_metrics/17_1_threshold_mismatches.sql [hunk 19]
```diff
@@ -0,0 +1,14 @@
+/* --title 'Threshold Mismatches' --width 3 */
+SELECT
+    main.recon_id,
+    CONCAT(main.target_table.catalog, '.', main.target_table.schema, '.', main.target_table.table_name) AS target_table,
```
Comment (vijaypavann-db) [inline src/databricks/labs/remorph/resources/reconcile/dashboards/reconciliation_metrics/17_1_threshold_mismatches.sql:4]: same as above

**Review by vijaypavann-db — CHANGES_REQUESTED** [review]

File: src/databricks/labs/remorph/resources/reconcile/dashboards/failed_table_list.sql (hunk 1 above)
Comment (bishwajit-db) [reply src/databricks/labs/remorph/resources/reconcile/dashboards/failed_table_list.sql]: New files were added

File: src/databricks/labs/remorph/resources/reconcile/dashboards/recon_details_pivot.sql (hunk 2 above)
Comment (bishwajit-db) [reply src/databricks/labs/remorph/resources/reconcile/dashboards/recon_details_pivot.sql]: New files were added

**Review by vijaypavann-db — APPROVED** [review]
LGTM

**Review by nfx — APPROVED** [review]
lgtm

**Pull request merged**